core/src/main/java/hudson/util/RingBufferLogHandler.java

```diff
@@ -86,13 +74,33 @@ public static int getDefaultRingBufferSize() {
     }
 
     @Override
-    public synchronized void publish(LogRecord record) {
-        int len = records.length;
-        records[(start + size) % len] = new LogRecordRef(record);
-        if (size == len) {
-            start = (start + 1) % len;
-        } else {
-            size++;
+    public void publish(LogRecord record) {
+        if (record.getParameters() != null) {
+            try {
+                LogRecord clone = new LogRecord(record.getLevel(), dummyFormatter.formatMessage(record));
+                clone.setLoggerName(record.getLoggerName());
+                clone.setMillis(record.getMillis());
+                clone.setSequenceNumber(record.getSequenceNumber());
+                clone.setSourceClassName(record.getSourceClassName());
+                clone.setSourceMethodName(record.getSourceMethodName());
+                clone.setThreadID(record.getThreadID());
+                Throwable t = record.getThrown();
+                if (t != null) {
+                    clone.setThrown(new ProxyException(t));
```

Comment:
Just to make sure any large objects that might be held by the stack trace are not also leaked.